include/pmacc/math/complex/Bessel.tpp

```diff
@@ -0,0 +1,276 @@
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file Bessel.tpp
+ *
+ *  Reference: Implementation is derived from a C++ implementation of
+ *             complex Bessel functions from C. Bond (2003).
+ *
+ *  Original source downloaded from: http://www.crbond.com
+ *  Download date: 2017/07/27
+ *  Files: CBESSJY.CPP, BESSEL.H
+ *  File-Header:
+ *      cbessjy.cpp -- complex Bessel functions.
+ *      Algorithms and coefficient values from "Computation of Special
+ *      Functions", Zhang and Jin, John Wiley and Sons, 1996.
+ *
+ *     (C) 2003, C. Bond. All rights reserved.
+ *
+ *  The website (http://www.crbond.com) furthermore states:
+ *  "This website contains a variety of materials related to
+ *  technology and engineering. Downloadable software, much of it
+ *  original, is available from some of the pages. All downloadable
+ *  software is offered freely and without restriction -- although
+ *  in most cases the files should be considered as works in progress
+ *  (alpha or beta level). Source code is also included for some
+ *  applications."
+ *
+ *  Code history:
+ *  1/03 -- Added C/C++ source files for real and complex gamma function
+ *  and psi function. Also added individual C/C++ files  for Bessel and
+ *  modified Bessel functions of 1st and 2nd kinds for real and complex
+ *  arguments. Updated Butterworth and Bessel filter tables with files of
+ *   extended parameters including polynomials, poles and component values.
+ *  6/04 -- Revised bessel.zip to correct errors in complex Bessel functions.
+ *
+ *  Further (re-)implementation of this code has been done by FZ Juelich.
+ *  URL: http://apps.jcns.fz-juelich.de/redmine/issues/569#change-2056
+ *  Above URL also includes a accuracy test report of this code against
+ *  SLATEC, MAPLE and MATHEMATICA.
+ */
+
+#pragma once
+
+#include "pmacc/types.hpp"
+#include "pmacc/algorithms/math.hpp"
+#include "pmacc/algorithms/TypeCast.hpp"
+#include "pmacc/math/Complex.hpp"
+
+#include <cmath>
+
+
+namespace pmacc
+{
+namespace algorithms
+{
+namespace math
+{
+namespace bessel
+{
+    namespace pmMath = pmacc::algorithms::math;
+
+    template<
+        typename T_Type,
+        typename T_TableA,
+        typename T_TableB,
+        typename T_TableA1,
+        typename T_TableB1
+    >
+    struct Cbesselj0Base
+    {
+        using result = pmacc::math::Complex< T_Type >;
+        using complex_T = pmacc::math::Complex< T_Type >;
+        using float_T = T_Type;
+
+        HDINLINE result operator( )( complex_T const & z )
+        {
+            T_TableA a;
+            T_TableB b;
+            T_TableA1 a1;
+            T_TableB1 b1;
+            result cj0;
+            T_Type const eps = float_T( 1.0e-15 );
```

Comment:
if this should be epsilon of double or float please use http://en.cppreference.com/w/cpp/types/numeric_limits/epsilon